src/test/java/com/uid2/shared/attest/UidOptOutClientTest.java

```diff
@@ -1,35 +1,104 @@
 import java.net.Proxy;
-import java.net.http.HttpClient;
+import java.net.http.HttpResponse;
 
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
 
 public class UidOptOutClientTest {
     private Proxy proxy = CloudUtils.defaultProxy;
-    private AttestationTokenRetriever mockAttestationTokenRetriever = mock(AttestationTokenRetriever.class);
+    private AttestationResponseHandler mockAttestationResponseHandler = mock(AttestationResponseHandler.class);
     private URLConnectionHttpClient mockHttpClient = mock(URLConnectionHttpClient.class);
     private UidOptOutClient optOutClient;
 
+    private MockedStatic<LoggerFactory> mockedLoggerFactory;
+    private static Logger mockedLogger = mock(Logger.class);
+
+    @BeforeClass
+    public static void classSetup() {
+
+    }
+
+    @AfterEach
+    public void close() {
+        mockedLoggerFactory.close();
+    }
+
     @BeforeEach
     void setUp() {
-        when(mockAttestationTokenRetriever.getOptOutJWT()).thenReturn("optOutJWT");
-        when(mockAttestationTokenRetriever.getCoreJWT()).thenReturn("coreJWT");
+        mockedLoggerFactory = mockStatic(LoggerFactory.class);
+        mockedLoggerFactory.when(() -> LoggerFactory.getLogger(any(Class.class))).thenReturn(mockedLogger);
+
+        when(mockAttestationResponseHandler.getOptOutJWT()).thenReturn("optOutJWT");
+        when(mockAttestationResponseHandler.getCoreJWT()).thenReturn("coreJWT");
 
         optOutClient = new UidOptOutClient(
                 "userToken", proxy,
-                mockAttestationTokenRetriever, mockHttpClient);
+                mockAttestationResponseHandler, mockHttpClient);
     }
 
     @Test
     void getJwtReturnsOptOutToken() {
         Assertions.assertEquals("optOutJWT", this.optOutClient.getJWT());
     }
+
+    @ParameterizedTest
+    @CsvSource(value = {
+            "/path1/path2?var=3;https://core.example.com/path1/path2?var=3",
+            ";https://core.example.com",
+            "/;https://core.example.com/",
+    }, delimiter = ';')
+    void usesOptOutUrlForDownloadWithPath(String path, String expectedFullPath) throws CloudStorageException, IOException {
+        HttpResponse<String> mockHttpResponse = mock(HttpResponse.class);
+        String expectedResponseBody = "Hello, world!";
+        when(mockHttpResponse.body()).thenReturn(expectedResponseBody);
+        when(mockHttpResponse.statusCode()).thenReturn(200);
+        when(mockHttpClient.get(any(), any())).thenReturn(mockHttpResponse);
+
+        when(mockAttestationResponseHandler.attested()).thenReturn(true);
+        when(mockAttestationResponseHandler.getAttestationToken()).thenReturn("testAttToken");
+        when(mockAttestationResponseHandler.getOptOutUrl()).thenReturn("https://core.example.com");
+
+        InputStream is = this.optOutClient.download(path);
+        verify(mockAttestationResponseHandler, atLeastOnce()).getOptOutUrl();

```

Comment:
why is it `atLeastOnce()` here?

Reply:
It gets called twice, once to check not null, then to get the value.

Reply:
Should we change `atLeastOnce()` to `times(2)`